src/py21cmfast/src/ps.c

```diff
@@ -2217,27 +2252,125 @@ float GaussLegendreQuad_Nion_MINI(int Type, int n, float growthf, float M2, floa
+      MassBin = (int)floor( (LogMass - MinMass )*inv_mass_bin_width );
+      MassBinLow = MinMass + mass_bin_width*(double)MassBin;
+      double sigmaM2 = Sigma_InterpTable[MassBin] + ( LogMass - MassBinLow )*( Sigma_InterpTable[MassBin+1] - Sigma_InterpTable[MassBin] )*inv_mass_bin_width;
+      nu_higher_limit = (delta1-delta2)*(delta1-delta2)/(sigmaM2*sigmaM2-sigma2*sigma2)/growthf/growthf;
+
+
+      //note we keep nupivot1 just in case very negative delta makes it reach that nu
+      LogMass = log(1.5e9); //jbm could be done outside and it'd be even faster
+      int MassBinpivot = (int)floor( (LogMass - MinMass )*inv_mass_bin_width );
+      double MassBinLowpivot = MinMass + mass_bin_width*(double)MassBinpivot;
+      double sigmapivot1 = Sigma_InterpTable[MassBinpivot] + ( LogMass - MassBinLowpivot )*( Sigma_InterpTable[MassBinpivot+1] - Sigma_InterpTable[MassBinpivot] )*inv_mass_bin_width;
+      double nupivot1 = (delta1-delta2)*(delta1-delta2)/(sigmapivot1*sigmapivot1)/growthf/growthf; //note, it does not have the sigma2 on purpose.
+
+      LogMass = log(5.3e5); //jbm could be done outside and it'd be even faster
+      MassBinpivot = (int)floor( (LogMass - MinMass )*inv_mass_bin_width );
+      MassBinLowpivot = MinMass + mass_bin_width*(double)MassBinpivot;
+      double sigmapivot2 = Sigma_InterpTable[MassBinpivot] + ( LogMass - MassBinLowpivot )*( Sigma_InterpTable[MassBinpivot+1] - Sigma_InterpTable[MassBinpivot] )*inv_mass_bin_width;
+      double nupivot2 = (delta1-delta2)*(delta1-delta2)/(sigmapivot2*sigmapivot2)/growthf/growthf;
+
+
+      double beta1 = (Alpha_star+Alpha_esc) * 9.0 * (0.5); //exponent for Fcollapprox for nu>nupivot1 (large M)
+      double beta2 = (Alpha_star+Alpha_esc) * 13.6 * (0.5); //exponent for Fcollapprox for nupivot1>nu>nupivot2 (small M)
+      double beta3 = (Alpha_star+Alpha_esc) * 21.0 * (0.5); //exponent for Fcollapprox for nu<nupivot2 (smallest M)
+      //beta2 fixed by continuity.
+
+
+      // // 3PLs
+      double fcollres=0.0;
+      double fcollres_high=0.0; //for the higher threshold to subtract
+
+
+      // re-written for further speedups
+      if (nu_higher_limit <= nupivot2){ //if both are below pivot2 don't bother adding and subtracting the high contribution
+      fcollres=(Fcollapprox(nu_lower_limit,beta3))*pow(nupivot2,-beta3);
+      fcollres_high=(Fcollapprox(nu_higher_limit,beta3))*pow(nupivot2,-beta3);
+      }
+      else {
+      fcollres_high=(Fcollapprox(nu_higher_limit,beta2))*pow(nupivot1,-beta2);
+      if (nu_lower_limit > nupivot2){
+        fcollres=(Fcollapprox(nu_lower_limit,beta2))*pow(nupivot1,-beta2);
+      }
+      else {
+        fcollres=(Fcollapprox(nupivot2,beta2))*pow(nupivot1,-beta2);
+        fcollres+=(Fcollapprox(nu_lower_limit,beta3)-Fcollapprox(nupivot2,beta3) )*pow(nupivot2,-beta3);
+      }
+
+      }
+
+      if (fcollres < fcollres_high){
+      return 1e-40;
+      }
+      return (fcollres-fcollres_high);
```

Comment:
hmm... so effectively, we also have a hard cut for the upper limit...

Reply:
Yes, we need hard cut on both ends